### PR TITLE
feat: add package names api for gateway

### PR DESCRIPTION
**Pull request by nichmor**
### Description

   
   will be used by new gateway for pixi search

### CODE REVIEW

File: crates/rattler_repodata_gateway/src/gateway/local_subdir.rs [hunk 1]
```diff
@@ -64,4 +64,9 @@ impl SubdirClient for LocalSubdirClient {
         })
         .await
     }
+
+    fn package_names(&self) -> Arc<[String]> {
```
Comment (baszalmstra) [inline crates/rattler_repodata_gateway/src/gateway/local_subdir.rs:68]: You are currently not caching this method so there is not need to return an `Arc`. Just use `Vec<String>`.

File: crates/rattler_repodata_gateway/src/gateway/query.rs [hunk 2]
```diff
@@ -271,11 +271,163 @@ impl GatewayQuery {
...
     fn into_future(self) -> Self::IntoFuture {
         self.execute().boxed()
     }
 }
+
+/// Represents a query for package names to execute with a [`Gateway`].
+///
+/// When executed the query will asynchronously load the package names from all
+/// subdirectories (combination of channels and platforms).
+///
```
Comment (baszalmstra) [inline crates/rattler_repodata_gateway/src/gateway/query.rs:287]:
```suggestion
```

File: crates/rattler_repodata_gateway/src/gateway/query.rs [hunk 3]
```diff
@@ -271,11 +271,163 @@ impl GatewayQuery {
...
+///
+/// Repodata is cached by the [`Gateway`] so executing the same query twice
+/// with the same channels will not result in the repodata being fetched
+/// twice.
```
Comment (baszalmstra) [inline crates/rattler_repodata_gateway/src/gateway/query.rs:291]: This doesnt apply to this function.

File: crates/rattler_repodata_gateway/src/gateway/query.rs [hunk 4]
```diff
@@ -271,11 +271,163 @@ impl GatewayQuery {
...
+    /// query. This allows reporting progress back to a user.
+    pub fn with_reporter(self, reporter: impl Reporter + 'static) -> Self {
+        Self {
+            reporter: Some(Arc::new(reporter)),
+            ..self
+        }
+    }
+
+    /// Execute the query and return the package names.
+    pub async fn execute(self) -> Result<Vec<String>, GatewayError> {
```
Comment (baszalmstra) [inline crates/rattler_repodata_gateway/src/gateway/query.rs:336]: This function is way more complex than it needs to be. There is no recursion in the names so you can do all of this concurrently. 

In pseudo code:

```
let package_names = Set::new();
for (channel, platform) in channels_and_platforms {
   subdir = get_or_create_subdir(..).await
   package_names.extend(subdir.package_names().await)
}

return package_names
```

You just need to do this concurrently for each `subdir`. Ill let you figure out how ;)

**Review by baszalmstra — CHANGES_REQUESTED** [review]

File: crates/rattler_repodata_gateway/src/gateway/query.rs (hunk 4 above)
Comment (nichmor) [reply crates/rattler_repodata_gateway/src/gateway/query.rs:331]: I've removed all unnecessary parts and just used FuturesUnordered to poll results.

Let me know if it's ok

**Review by baszalmstra — COMMENTED** [review]
lets return `PackageName` as the type. Otherwise looks great!

**Comment by nichmor** [comment]
> lets return `PackageName` as the type. Otherwise looks great!

I've changed NameQuery to return PackageName.

I wasn't sure if the SubDirClient trait should return PackageName instead of strings so I've decided to keep it more simple building block.

**Review by baszalmstra — APPROVED** [review]

**Pull request merged**